### PR TITLE
Support bfd management

### PR DESCRIPTION
**Pull request by bobz965**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

[15-kube-ovn-ecmp-with-bfd.md](https://github.com/kubeovn/kube-ovn/files/10824869/15-kube-ovn-ecmp-with-bfd.md)

### CODE REVIEW

File: pkg/controller/subnet.go [hunk 1]
```diff
@@ -941,9 +940,20 @@ func (c *Controller) reconcileSubnet(subnet *kubeovnv1.Subnet) error {
...
 
-	if err := c.reconcileOvnRoute(subnet); err != nil {
-		klog.Errorf("reconcile OVN route for subnet %s failed: %v", subnet.Name, err)
-		return err
+	if subnet.Spec.Vpc == util.DefaultVpc {
+		if err := c.reconcileOvnDefaultVpcRoute(subnet); err != nil {
+			klog.Errorf("reconcile default vpc ovn route for subnet %s failed: %v", subnet.Name, err)
+			return err
+		}
+		return nil
```
Comment (hongzhen-ma) [inline pkg/controller/subnet.go:948]: 不能 return nil，还需要继续后面的处理

File: pkg/controller/subnet.go [hunk 2]
```diff
@@ -941,9 +940,20 @@ func (c *Controller) reconcileSubnet(subnet *kubeovnv1.Subnet) error {
...
+	}
+
+	if subnet.Spec.Vpc != util.DefaultVpc {
+		if err := c.reconcileOvnCustomVpcRoute(subnet); err != nil {
+			klog.Errorf("reconcile custom vpc ovn route for subnet %s failed: %v", subnet.Name, err)
+			return err
+		}
+		return nil
```
Comment (hongzhen-ma) [inline pkg/controller/subnet.go:956]: 同样，不能 return nil

File: pkg/controller/subnet.go [hunk 3]
```diff
@@ -1079,11 +1089,226 @@ func (c *Controller) reconcileNamespaces(subnet *kubeovnv1.Subnet) error {
...
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+		klog.Errorf("failed to get vpc %s, %v", vpcName, err)
+		return err
+	}
+	lrpEipName := fmt.Sprintf("%s-%s", vpcName, c.config.ExternalGatewaySwitch)
+	lrpEip, err := c.config.KubeOvnClient.KubeovnV1().OvnEips().Get(context.Background(), lrpEipName, metav1.GetOptions{})
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
```
Comment (hongzhen-ma) [inline pkg/controller/subnet.go:1123]: add log

File: pkg/controller/subnet.go [hunk 4]
```diff
@@ -1079,11 +1089,226 @@ func (c *Controller) reconcileNamespaces(subnet *kubeovnv1.Subnet) error {
...
+	// dst 0.0.0.0 nexthop external switch gw
+	// allow all subnet access external based snat
+	// also, this will not conflict with policy route
+
+	defualtExternalSubnet, err := c.subnetsLister.Get(c.config.ExternalGatewaySwitch)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			klog.Error("failed to get default external switch subnet %s: %v", c.config.ExternalGatewaySwitch)
+			return err
+		}
```
Comment (hongzhen-ma) [inline pkg/controller/subnet.go:1195]: not  found 的情况，也需要处理

File: pkg/controller/subnet.go [hunk 5]
```diff
@@ -1079,11 +1089,226 @@ func (c *Controller) reconcileNamespaces(subnet *kubeovnv1.Subnet) error {
...
+
+func (c *Controller) reconcileVpcDelNormalStaticRoute(vpcName string) error {
+	// normal static route is prior than ecmp bfd static route
+	// if use ecmp bfd static route, normal static route should not exist
+	defualtExternalSubnet, err := c.subnetsLister.Get(c.config.ExternalGatewaySwitch)
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			klog.Error("failed to get default external switch subnet %s: %v", c.config.ExternalGatewaySwitch)
+			return err
+		}
```
Comment (hongzhen-ma) [inline pkg/controller/subnet.go:1271]: not found 情况

File: pkg/controller/subnet.go [hunk 6]
```diff
@@ -1476,6 +1703,32 @@ func (c *Controller) reconcileOvnRoute(subnet *kubeovnv1.Subnet) error {
...
+		return err
+	}
+
+	if vpc.Spec.EnableExternal && vpc.Spec.EnableBfd && subnet.Spec.EnableEcmp {
+		klog.V(3).Infof("add bfd and external static ecmp route for vpc %s, subnet %s", vpc.Name, subnet.Name)
+		// handle vpc static route
+		// use static ecmp route with bfd
+		// bfd ecmp static route depend on subnet cidr
+		if err := c.reconcileVpcUseBfdStaticRoute(vpc.Name, subnet.Name); err != nil {
+			klog.Errorf("failed to reconcile vpc %q bfd static route", vpc.Name)
```
Comment (hongzhen-ma) [inline pkg/controller/subnet.go:1726]: return err ?

File: pkg/controller/vpc.go [hunk 7]
```diff
@@ -508,10 +527,39 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
+			if err := c.handleAddVpcExternal(key); err != nil {
+				klog.Errorf("failed to add external connection for vpc %s, error %v", key, err)
+				return err
+			}
+		}
+		if vpc.Spec.EnableBfd {
+			klog.V(3).Infof("remove normal static ecmp route for vpc %s", vpc.Name)
+			// auto remove normal type static route, if use ecmp based bfd
+			if err := c.reconcileVpcDelNormalStaticRoute(vpc.Name); err != nil {
+				klog.Errorf("failed to reconcile del vpc %q normal static route", vpc.Name)
```
Comment (hongzhen-ma) [inline pkg/controller/vpc.go:543]: return err ?

File: pkg/controller/vpc.go [hunk 8]
```diff
@@ -508,10 +527,39 @@ func (c *Controller) handleAddOrUpdateVpc(key string) error {
...
+			}
+		}
+		if !vpc.Spec.EnableBfd {
+			klog.V(3).Infof("add normal static ecmp route for vpc %s", vpc.Name)
+			// auto add normal type static route, if not use ecmp based bfd
+			if err := c.reconcileVpcAddNormalStaticRoute(vpc.Name); err != nil {
+				klog.Errorf("failed to reconcile vpc %q bfd static route", vpc.Name)
```
Comment (hongzhen-ma) [inline pkg/controller/vpc.go:550]: return err?

File: pkg/ovs/ovn-nbctl-legacy.go [hunk 9]
```diff
@@ -1162,11 +1185,20 @@ func parseLrRouteListOutput(output string) (routeList []*StaticRoute, err error)
...
+				Policy:  fields[2],
+				CIDR:    fields[0],
+				NextHop: fields[1],
+			})
+		} else {
+			routeList = append(routeList, &StaticRoute{
+				Policy:  fields[2],
+				CIDR:    fields[0],
+				NextHop: fields[1],
+				ECMP:    fields[4],
```
Comment (hongzhen-ma) [inline pkg/ovs/ovn-nbctl-legacy.go:1199]: 4 or 3 ?

File: pkg/ovs/ovn-nbctl-legacy.go [hunk 10]
```diff
@@ -3158,3 +3190,93 @@ func (c *LegacyClient) GetNatIPInfo(uuid string) (string, error) {
...
+// CreateBfd find ovn bfd uuid by lrp name and dst ip
+func (c LegacyClient) CreateBfd(lrp, dstIp string, minTx, minRx, detectMult int) (string, error) {
+	var err error
+	var output string
+	if bfdId, err := c.FindBfd(lrp, dstIp); err == nil {
+		return bfdId, nil
+	}
+	if output, err = c.ovnNbCommand("create", "bfd", fmt.Sprintf("logical_port=%s", lrp), fmt.Sprintf("dst_ip=%s", dstIp),
+		fmt.Sprintf("min_tx=%d ", minTx), fmt.Sprintf("min_rx=%d", minRx), fmt.Sprintf("detect_mult=%d", detectMult)); err != nil {
+		klog.Errorf("faild to create bfd for lrp %q dst ip %q, output %s, %v", lrp, dstIp, output, err)
```
Comment (hongzhen-ma) [inline pkg/ovs/ovn-nbctl-legacy.go:3242]: %q ?

**Review by hongzhen-ma — CHANGES_REQUESTED** [review]

File: pkg/controller/config.go [hunk 11]
```diff
@@ -161,6 +165,10 @@ func ParseFlags() (*Configuration, error) {
 
 		argGCInterval      = pflag.Int("gc-interval", 360, "The interval between GC processes, default 360 seconds")
 		argInspectInterval = pflag.Int("inspect-interval", 20, "The interval between inspect processes, default 20 seconds")
+
+		argBfdMinTx      = pflag.Int("bfd-min-tx", 100, "This is the minimum interval, in milliseconds, ovn would like to use when transmitting BFD Control packets")
+		argBfdMinRx      = pflag.Int("bfd-min-rx", 100, "This is the minimum interval, in milliseconds, between received BFD Control packets")
+		argBfdDetectMult = pflag.Int("detect-mult", 3, "The negotiated transmit interval, multiplied by this value, provides the Detection Time for thereceiving system in Asynchronous mode.")
```
[the remark this answers is not in the document]
Comment (bobz965) [reply pkg/controller/config.go:171]: ok

File: pkg/controller/subnet.go (hunk 1 above)
Comment (bobz965) [reply pkg/controller/subnet.go:948]: ok

File: pkg/controller/subnet.go (hunk 2 above)
Comment (bobz965) [reply pkg/controller/subnet.go:956]: ok

File: pkg/controller/subnet.go (hunk 3 above)
Comment (bobz965) [reply pkg/controller/subnet.go:1123]: ok

File: pkg/controller/subnet.go (hunk 4 above)
Comment (bobz965) [reply pkg/controller/subnet.go:1195]: ok

File: pkg/controller/subnet.go (hunk 6 above)
Comment (bobz965) [reply pkg/controller/subnet.go:1726]: ok

File: pkg/controller/vpc.go (hunk 7 above)
Comment (bobz965) [reply pkg/controller/vpc.go:543]: ok

File: pkg/controller/vpc.go (hunk 8 above)
Comment (bobz965) [reply pkg/controller/vpc.go:550]: ok

File: pkg/ovs/ovn-nbctl-legacy.go (hunk 9 above)
Comment (bobz965) [reply pkg/ovs/ovn-nbctl-legacy.go:1199]: 是 4

[root@pc-node-1 03-cust-vpc]# k ko nbctl lr-route-list vpc1
IPv4 Routes
Route Table <main>:
           192.168.0.0/24              10.5.204.104 src-ip ecmp ecmp-symmetric-reply bfd
           192.168.0.0/24              10.5.204.114 src-ip ecmp ecmp-symmetric-reply bfd
           192.168.0.0/24              10.5.204.124 src-ip ecmp ecmp-symmetric-reply bfd

File: pkg/ovs/ovn-nbctl-legacy.go (hunk 10 above)
Comment (bobz965) [reply pkg/ovs/ovn-nbctl-legacy.go:3242]: ok

**Review by hongzhen-ma — APPROVED** [review]

**Pull request merged**